### DISCUSSION TITLE
Can't import named exports from an external NPM library with TypeScript

### DISCUSSION THREAD

**Discussion by malikalimoekhamedov** [vercel/next.js#11631] [Help]
Hello friends,

I'm unable to import named exports from an external component library I've created as it is described in the [official documentation](https://nextjs.org/docs/advanced-features/dynamic-import). The situation seems to be very niche as I struggle to find anything online that would bring me closer to the solution. Did anyone manage to get it to work?

Here comes the context...

### Component

**Modal component** (fragment)
```ts
const Modal: React.FC<ModalProps> = ({ onCancelCallback, children }) => {...};
export default Modal;
```
As you can see, I export it as a default from Modal.tsx but later re-export it as a named export in every folder's index.ts file like so:
```ts
export { default as Modal } from './Modal';
```
then
```ts
export { Modal } from './Modal';
```
and, finally:
```ts
export * from './components';
```

### Interface
**ModalProps**
```ts
export interface ModalProps {
  onCancelCallback?: () => void;
}
```
## In Next.js
Here's how I'm importing my external "Modal" component in one of the Next.js components (not pages):

**nextjscomponent.tsx**
```ts
import dynamic from 'next/dynamic';
const Modal = dynamic(() => import('my-ui-kit').then((mod) => mod.Modal), {
  ssr: false,
});
```

**TypeScript error**
```sh
Argument of type '() => Promise<React.ComponentClass<never, any> | React.FunctionComponent<never> | { default: React.ComponentType<never>; } | React.FC<ModalProps>>' is not assignable to parameter of type 'DynamicOptions<{}> | (() => LoaderComponent<{}>) | LoaderComponent<{}>'.
  Type '() => Promise<React.ComponentClass<never, any> | React.FunctionComponent<never> | { default: React.ComponentType<never>; } | React.FC<ModalProps>>' is not assignable to type '() => LoaderComponent<{}>'.
    Type 'Promise<ComponentClass<never, any> | FunctionComponent<never> | { default: ComponentType<never>; } | FC<ModalProps>>' is not assignable to type 'LoaderComponent<{}>'.
      Type 'ComponentClass<never, any> | FunctionComponent<never> | { default: ComponentType<never>; } | FC<ModalProps>' is not assignable to type 'ComponentClass<{}, any> | FunctionComponent<{}> | { default: ComponentType<{}>; }'.
        Type 'ComponentClass<never, any>' is not assignable to type 'ComponentClass<{}, any> | FunctionComponent<{}> | { default: ComponentType<{}>; }'.
          Type 'ComponentClass<never, any>' is not assignable to type 'ComponentClass<{}, any>'.
            Types of property 'getDerivedStateFromProps' are incompatible.
              Type 'GetDerivedStateFromProps<never, any> | undefined' is not assignable to type 'GetDerivedStateFromProps<{}, any> | undefined'.
                Type 'GetDerivedStateFromProps<never, any>' is not assignable to type 'GetDerivedStateFromProps<{}, any>'.
                  Types of parameters 'nextProps' and 'nextProps' are incompatible.
                    Type 'Readonly<{}>' is not assignable to type 'never'.ts(2345)
```

## Notes
- I'm using Rollup to transpile the contents of "src" folder with component and index files into "dist" folder which ends up with index.cjs.js (for CommonJS), index.esm.js (for ES modules) and a bunch of automatically generated .d.ts files.
- I'm using Yarn link to test the integration of my external library into my Next.js project locally.

Any help is highly appreciated. Thank you in advance.

**Comment by wereHamster** [comment]
It seems that you have two different versions of @types/react installed. You should only have a single instance of that package in your node_modules folder.

**Reply by malikalimoekhamedov** [reply]
@wereHamster, thanks for chipping in.

So, I rand a quick "yarn why @types/react" and saw the following:
```sh
yarn why v1.22.4
[1/4] 🤔  Why do we have the module "@types/react"...?
[2/4] 🚚  Initialising dependency graph...
[3/4] 🔍  Finding dependency...
[4/4] 🚡  Calculating file sizes...
=> Found "@types/react@16.9.32"
info Has been hoisted to "@types/react"
info This module exists because it's specified in "devDependencies".
info Disk size without dependencies: "164KB"
info Disk size with unique dependencies: "1.98MB"
info Disk size with transitive dependencies: "1.98MB"
info Number of shared dependencies: 2
=> Found "@types/react-dom#@types/react@16.9.31"
info This module exists because "@types#react-dom" depends on it.
info Disk size without dependencies: "164KB"
info Disk size with unique dependencies: "1.98MB"
info Disk size with transitive dependencies: "1.98MB"
info Number of shared dependencies: 2
✨  Done in 0.99s.
```
I then tried three things:
1. Unlink my library > yarn cache clean > reinstall the library (this time from a freshly generated tgz tarball). The issue persisted
2. Yarn remove @types/react from the Next.js folder (effectively leaving only @types/react v16.9.31 hoisted from @types/react-dom). The issue persisted
3. Completely removed node_modules and yarn.lock and reinstalled all packages (including my library from the tarball). The issue persisted.

I'm still seing the same error message. **Where would you look next?**

By the way, that same component works fine with dynamic loading when imported from Next.js project's own components folder but the goal is to use the external UI kit, obviously.

**Comment by malikalimoekhamedov** [comment]
Friends, any ideas with regards to this eventually?

**Reply by r0skar** [reply]
You can pass the props type of the dynamic component as a generic or as a last resort you can pass `any`:

```typescript
const Modal = dynamic<any>(() => import('my-ui-kit').then((mod) => mod.Modal), {
  ssr: false,
});
```

**Reply by malikalimoekhamedov** [reply]
@r0skar, brilliant! I feel like this aspect should be added to the Next.js official docs for folks using TypeScript.

I've done the following:

**In my external ui kit**
```ts
export type { ModalProps } from './Modal.interface';
```

**In Next.js**
```ts
import { ModalProps } from 'my-ui-kit';
const Modal = dynamic<ModalProps>(() => import('my-ui-kit').then((mod) => mod.Modal), {
  ssr: false,
});
```

This made TypeScript (and myself) happy.

---

I guess this solves this particular issue, but now I run into a different (albeit related) issue because I'm using React hooks inside this Modal. Here's the error I'm getting:

```sh
Error: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
See https://fb.me/react-invalid-hook-call for tips about how to debug and fix this problem.
```
Just like in the case of dynamic import, using React hooks inside a component wasn't an issue for Next.js when this component was internal to Next.js project and was imported (untranspiled) from its components folder.

Right now, Next understands my modal as follows:
```ts
const Modal: React.ComponentType<ModalProps>
```

Is there a standard "Next.js recommended" procedure to deal with this kind of issues?

Thanks you in advance. Made my day!

**Comment by davidye** [comment]
I'm also seeing this issue with named exports only - default type checks just fine.
